src/geos.c

```diff
@@ -896,12 +897,22 @@ GEOSCoordSequence* coordseq_from_buffer(GEOSContextHandle_t ctx, const double* b
     if ((cs1 == dims * 8) && (cs2 == 8)) {
       /* C-contiguous memory */
       int hasZ = dims == 3;
+      if (dims > 3) {
+        snprintf(last_error, 1024,
+                 "IllegalArgumentException: Ordinate dimension cannot be larger than 3");
+        return NULL;
```

Comment:
This is a bit an ugly solution (the usage of `last_error` to set a GEOS exception message on our side, and the need to pass through `last_error` in the function). 

But so the problem here is that this function needs to return NULL to signal an error (the way it is currently written), and inside the ufunc any NULL (it can also come from GEOS itself) is considered as a "GEOSException". So if I don't set any `last_error`, you get a "blank" GEOSException error.

I first tried to raise a Python error, but, then we would need to pass through the `errstate` instead, and setting a python error string (eg `PyErr_SetString`) also requires the GIL, which is not the case within this function. 

One alternative that is maybe a bit cleaner is to check for the dimension one level up in ufuncs.c (then we don't need to passthrough `last_error`, as that will be available there anyway). Or that would maybe even allow to make it a python error.

Reply:
I see the issue, I am fine with this solution actually, but maybe you could consider an alternative: let this function return an `errstate` and put the `CoordinateSequence` pointer as an additional argument. That would of course involve a minor change in the places where this function is called.

This specific error case should be registered as an additional enum value with corresponding error raising in the `GEOS_FINISH` macro.

This pattern is used in some GEOS functions that return doubles. Also I tended to use it in some custom PyGEOS functions.

Reply:
Changing it to return the errstate and registering the case in the enum and handling it in GEOS_FINISH sounds good (I suppose with setting it as a Python error then?). 

Now, thinking about it a bit more, you get a bit an inconsistent state where depending on the memory layout of the input array you would get a GEOSException or a ValueError if you have >3 ordinate dimension. 
It might be nice to always get a consistent error for that case. And so if we are handling this with a new errstate enum, I could also set this inside the ufunc code, _before_ calling `coordseq_from_buffer` (instead of checking it inside `coordseq_from_buffer`).

Reply:
Sounds better! That would be an easy fix, if you do it before releasing the GIL then you can just directly set the PyErr.

In the cython code this is already checked.

Reply:
OK, updated with this much simpler solution. 

I also added a check that the dimension is at least 2, because right now you could actually create a LineString/LinearRing with a 1D ordinate dimension, which then gets filled with zeros for the y dimension:

```python
In [2]: shapely.linestrings(np.random.randn(3, 2, 1))
Out[2]: 
array([<shapely.geometry.LineString LINESTRING (0.9497887286261655 0, -0.3278272663692793 0)>,
       <shapely.geometry.LineString LINESTRING (1.9726398481813794 0, 0.8734770826550462 0)>,
       <shapely.geometry.LineString LINESTRING (1.054188743128901 0, -1.0106490603521021 0)>],
      dtype=object)
```

I didn't use the exact same error message as the cython one, because here it needs to be more general (in the cython code the input is always a 2D array).